internal/utils/crossnamespace/kongreferencegrant.go

```diff
@@ -0,0 +1,172 @@
+package crossnamespace
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	configurationv1alpha1 "github.com/kong/kong-operator/api/configuration/v1alpha1"
+)
+
+// ErrCrossNamespaceReferenceNotGranted is an error type that indicates a cross-namespace
+// reference is not granted by any KongReferenceGrant.
+type ErrCrossNamespaceReferenceNotGranted struct {
+	FromNamespace string
+	FromGVK       metav1.GroupVersionKind
+	ToNamespace   string
+	ToName        string
+	ToGVK         metav1.GroupVersionKind
+}
+
+// Error returns a formatted error message indicating that a cross-namespace reference
+// is not permitted.
+func (e *ErrCrossNamespaceReferenceNotGranted) Error() string {
+	return fmt.Sprintf("cross-namespace reference to %s %s in namespace %s from %s.%s in namespace %s is not granted by any KongReferenceGrant",
+		e.ToGVK.Kind,
+		e.ToName,
+		e.ToNamespace,
+		e.FromGVK.Kind,
+		e.FromGVK.Group,
+		e.FromNamespace,
+	)
+}
+
+// IsCrossNamespaceReferenceNotGranted checks if the provided error is or wraps
+// an ErrCrossNamespaceReferenceNotGranted error, indicating that a cross-namespace
+// reference was attempted without the proper ReferenceGrant permissions.
+// It returns true if the error matches this type, false otherwise.
+func IsCrossNamespaceReferenceNotGranted(err error) bool {
+	var target *ErrCrossNamespaceReferenceNotGranted
+	return errors.As(err, &target)
+}
+
+// CheckKongReferenceGrantForResource verifies that a cross-namespace reference is permitted by checking
+// for an appropriate KongReferenceGrant. It validates whether a resource in one namespace (from) is allowed
+// to reference a resource in another namespace (to).
+//
+// Parameters:
+//   - cl: The Kubernetes client used to query KongReferenceGrant resources
+//   - ctx: The context for the operation
+//   - fromNamespace: The namespace containing the resource making the reference
+//   - toNamespace: The namespace containing the resource being referenced
+//   - toName: The name of the resource being referenced
+//   - fromGVK: The GroupVersionKind of the resource making the reference
+//   - toGVK: The GroupVersionKind of the resource being referenced
+//
+// Returns an error if:
+//   - The verification of the KongReferenceGrant fails
+//   - No valid KongReferenceGrant exists that permits the cross-namespace reference
+//
+// Returns nil if the cross-namespace reference is properly granted.
+func CheckKongReferenceGrantForResource(
+	cl client.Client,
+	ctx context.Context,
```

Comment:
Context should always be first.

```suggestion
	ctx context.Context,
	cl client.Client,
```